mcp/mcp-server/src/main/java/software/amazon/smithy/java/mcp/server/HttpMcpProxy.java

```diff
@@ -0,0 +1,185 @@
+            return this;
+        }
+
+        public HttpMcpProxy build() {
+            if (endpoint == null || endpoint.isEmpty()) {
+                throw new IllegalArgumentException("Endpoint must be provided");
+            }
+            return new HttpMcpProxy(this);
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public CompletableFuture<JsonRpcResponse> rpc(JsonRpcRequest request) {
+        try {
+            byte[] body = JSON_CODEC.serializeToString(request).getBytes(StandardCharsets.UTF_8);
+            LOG.trace("Sending HTTP request to {}", endpoint);
+
+            HttpRequest httpRequest = HttpRequest.builder()
+                    .uri(endpoint)
+                    .method("POST")
+                    .withAddedHeader("Content-Type", "application/json")
+                    .withAddedHeader("Accept", "application/json, text/event-stream")
+                    .body(DataStream.ofBytes(body, "application/json"))
+                    .build();
+
+            Context context = Context.create();
+            context.put(HttpContext.HTTP_REQUEST_TIMEOUT, timeout);
+
+            if (signer != null) {
+                httpRequest = signer.sign(httpRequest, null, context);
+            }
+
+            HttpResponse response = transport.send(context, httpRequest);
+            LOG.trace("Received HTTP response with status: {}", response.statusCode());
+
+            if (response.statusCode() < 200 || response.statusCode() >= 300) {
+                return CompletableFuture.completedFuture(handleErrorResponse(response));
+            }
+
+            return CompletableFuture.completedFuture(JsonRpcResponse.builder()
+                    .deserialize(JSON_CODEC.createDeserializer(response.body().asByteBuffer()))
+                    .build());
+        } catch (Exception e) {
+            return CompletableFuture.failedFuture(e);
+        }
+    }
+
+    private JsonRpcResponse handleErrorResponse(HttpResponse response) {
+        long contentLength = response.body().contentLength();
+        String errorMessage = "HTTP " + response.statusCode();
+
+        if (contentLength > 0) {
+            String contentType = response.body().contentType();
+            byte[] bodyBytes = ByteBufferUtils.getBytes(response.body().asByteBuffer());
+
+            if (contentType != null && contentType.contains("application/json")) {
+                try {
+                    return JsonRpcResponse.builder()
+                            .deserialize(JSON_CODEC.createDeserializer(bodyBytes))
+                            .build();
+                } catch (Exception e) {
+                    LOG.warn("Failed to deserialize JSON error response", e);
+                }
+            }
+
+            String errorBody = new String(bodyBytes, 0, Math.min(200, bodyBytes.length), StandardCharsets.UTF_8);
+            errorMessage = errorMessage + ": " + errorBody;
```

Comment:
Explicitly show that this is truncated if we are picking the first 200 bytes only.

Something like

```
int length = Math.min(200, bodyBytes.length)

errorMessage = errorMessage + ": " + errorBody + (length != bodyBytes.length ? " (truncated)" + "";
```